hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala

```diff
@@ -92,410 +72,535 @@ object LowerMatrixIR {
-            scans += (s -> x)
+            scanBuilder += (s -> x)
+            Ref(s, x.typ)
+          case x if IsAggResult(x) =>
+            assert(!ContainsScan(x))
+            val s = genUID()
+            aggBuilder += (s -> x)
             Ref(s, x.typ)
           case _ =>
-            MapIR(f)(ir)
+            MapIR(lift)(ir)
         }
 
-        val b0 = f(ir)
-
-        val b: IRProxy =
-          if (ContainsAgg(b0)) {
-            irRange(0, 'row (entriesField).len)
-              .filter('i ~> !'row (entriesField)('i).isNA)
-              .arrayAgg('i ~>
-                (aggLet(sa = 'global (colsField)('i),
-                  g = 'row (entriesField)('i))
-                  in b0))
-          } else
-            b0
+        val e = Env[IR]("global" -> SelectFields(Ref("global", loweredChild.typ.globalType),
+          child.typ.globalType.fieldNames))
+        val substEnv = BindingEnv(e, Some(e), Some(e))
+
+        var b0 = lift(Subst(newCol, substEnv))
+        val aggs = aggBuilder.result()
+        val scans = scanBuilder.result()
+
+        val idx = Ref(genUID(), TInt32())
+        val idxSym = Symbol(idx.name)
+
+        val aggTransformer: IRProxy => IRProxy = if (aggs.isEmpty)
+          identity
+        else {
+          val aggStruct = MakeStruct(aggs)
+          val aggResultArray = loweredChild.aggregate(
+            aggLet(va = 'row) {
+              irRange(0, 'global (colsField).len)
+                .aggElements('__element_idx, '__result_idx)(
+                  let(sa = 'global (colsField)('__result_idx)) {
+                    aggLet(sa = 'global (colsField)('__element_idx),
+                      g = 'row (entriesField)('__element_idx)) {
+                      aggFilter(!'g.isNA, aggStruct)
+                    }
+                  })
+            })
+          val ident = genUID()
+          ab += ((ident, aggResultArray))
 
-        scans.result().foldLeft(b) { case (acc, (s, x)) =>
-          (env: E) => {
-            Let(s, x, acc(env))
-          }
+          val aggResultRef = Ref(genUID(), aggResultArray.typ)
+          val aggResultElementRef = Ref(genUID(), aggResultArray.typ.asInstanceOf[TArray].elementType)
+
+          b0 = aggs.foldLeft[IR](b0) { case (acc, (name, _)) => Let(name, GetField(aggResultElementRef, name), acc) }
+          b0 = Let(aggResultElementRef.name, ArrayRef(aggResultRef, idx), b0)
+
+          x: IRProxy => let.applyDynamicNamed("apply")((aggResultRef.name, RelationalRef(ident, aggResultArray.typ))).apply(x)
         }
-      }
 
-      val lc = lower(child)
-      val e = Env[IR]("va" -> Ref("row", lc.typ.rowType),
-        "global" -> SelectFields(Ref("global", lc.typ.globalType), child.typ.globalType.fieldNames))
-      lc.mapRows(
-        liftScans(Subst(newRow, BindingEnv(e, scan = Some(e), agg = Some(e))))
-          .insertFields(entriesField -> 'row (entriesField)))
-    }
+        val scanTransformer: IRProxy => IRProxy = if (scans.nonEmpty) {
```

Comment:
This should flip to `if (scans.isEmpty)` to match `aggTransformer`.

Reply:
Addressed. Also added a test for a rebase collision bug I found.